v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraDMLGenerator.java

```diff
@@ -0,0 +1,438 @@
+   *     data.
+   *     <p>This method: 1. Extracts column values from the provided request using the
+   *     `getColumnValues` method. 2. Combines the column values with the primary key column values.
+   *     3. Constructs the upsert statement using the `getUpsertStatementCQL` method.
+   *     <p>The upsert statement ensures that the record is inserted or updated in the Spanner table
+   *     based on the primary key.
+   */
+  private static DMLGeneratorResponse generateUpsertStatement(
+      SpannerTable spannerTable,
+      SourceTable sourceTable,
+      DMLGeneratorRequest dmlGeneratorRequest,
+      Map<String, PreparedStatementValueObject<?>> pkColumnNameValues) {
+    Map<String, PreparedStatementValueObject<?>> columnNameValues =
+        getColumnValues(
+            spannerTable,
+            sourceTable,
+            dmlGeneratorRequest.getNewValuesJson(),
+            dmlGeneratorRequest.getKeyValuesJson(),
+            dmlGeneratorRequest.getSourceDbTimezoneOffset());
+    return getUpsertStatementCQL(
+        sourceTable.getName(),
+        Instant.now().toEpochMilli() * 1000,
+        columnNameValues,
+        pkColumnNameValues);
+  }
+
+  /**
+   * Constructs an upsert (insert or update) CQL statement for a Cassandra or similar database using
+   * the provided table name, timestamp, column values, and primary key values.
+   *
+   * @param tableName the name of the table to which the upsert statement applies.
+   * @param timestamp the timestamp (in microseconds) to use for the operation.
+   * @param columnNameValues a map of column names and their corresponding prepared statement value
+   *     objects for non-primary key columns.
+   * @param pkColumnNameValues a map of primary key column names and their corresponding prepared
+   *     statement value objects.
+   * @return a {@link DMLGeneratorResponse} containing the generated CQL statement and a list of
+   *     values to be used with the prepared statement.
+   *     <p>This method: 1. Iterates through the primary key and column values, appending column
+   *     names and placeholders to the generated CQL statement. 2. Constructs the `INSERT INTO` CQL
+   *     statement with the provided table name, columns, and placeholders. 3. Appends a `USING
+   *     TIMESTAMP` clause to include the provided timestamp in the statement. 4. Creates a list of
+   *     values to bind to the placeholders in the prepared statement.
+   *     <p>The returned response contains the complete prepared CQL statement and the values
+   *     required to execute it.
+   */
+  private static DMLGeneratorResponse getUpsertStatementCQL(
+      String tableName,
+      long timestamp,
+      Map<String, PreparedStatementValueObject<?>> columnNameValues,
+      Map<String, PreparedStatementValueObject<?>> pkColumnNameValues) {
+
+    String escapedTableName = "\"" + tableName.replace("\"", "\"\"") + "\"";
+    List<Map.Entry<String, PreparedStatementValueObject<?>>> allEntries =
```

Comment:
Do we need this intermediate map?
Please let us know if the below might work too:
Consider something on the lines of 
```java
// import com.google.common.collect.Sets; 

Sets.SetView<String> allColumnNamesView = Strem.concat(Sets.union(pkColumnNameValues.keySet(), columnNameValues.keySet()))
String allColumns =
allColumnNamesView.stream()
  .map(entry -> "\"" + entry.getKey().replace("\"", "\"\"") + "\"")
            .collect(Collectors.joining(", "));
            
String placeholders = allColumnNamesView.stream().map(entry -> "?").collect(Collectors.joining(", "))
```

The view is unmodifiable and lazy. It would have lesser memory footprint than the map, also it might be a slightly more readable option to a list of entry set.